web/text_layer_builder.js

```diff
@@ -83,37 +91,83 @@ class TextLayerBuilder {
-
-    const textLayerFrag = document.createDocumentFragment();
-    this.textLayerRenderTask = renderTextLayer({
-      textContent: this.textContent,
-      textContentStream: this.textContentStream,
-      container: textLayerFrag,
-      viewport: this.viewport,
-      textDivs: this.textDivs,
-      textContentItemsStr: this.textContentItemsStr,
-      timeout,
-    });
-    this.textLayerRenderTask.promise.then(
-      () => {
-        this.textLayerDiv.append(textLayerFrag);
-        this.#finishRendering();
-        this.highlighter?.enable();
-        this.accessibilityManager?.enable();
-      },
-      function (reason) {
-        // Cancelled or failed to render text layer; skipping errors.
+    const scale = viewport.scale * (globalThis.devicePixelRatio || 1);
+
+    if (this.renderingDone) {
+      this.#update(viewport);
+      if (scale !== this.scale) {
+        updateTextLayer({
+          viewport,
+          scale,
+          textDivs: this.textDivs,
+          textDivProperties: this.textDivProperties,
+          isOffscreenCanvasSupported: this.isOffscreenCanvasSupported,
+        });
+        this.scale = scale;
       }
-    );
+      this.show();
+    } else {
+      this.cancel();
+
+      this.highlighter?.setTextMapping(this.textDivs, this.textContentItemsStr);
+      this.accessibilityManager?.setTextMapping(this.textDivs);
+
+      const textLayerFrag = document.createDocumentFragment();
+      this.textLayerRenderTask = renderTextLayer({
+        textContent: this.textContent,
+        textContentStream: this.textContentStream,
+        container: textLayerFrag,
+        viewport,
+        textDivs: this.textDivs,
+        textDivProperties: this.textDivProperties,
+        textContentItemsStr: this.textContentItemsStr,
+        isOffscreenCanvasSupported: this.isOffscreenCanvasSupported,
+        timeout,
+        scale,
+      });
+      this.textLayerRenderTask.promise.then(
+        () => {
+          this.div.append(textLayerFrag);
+          this.#finishRendering();
+          this.highlighter?.enable();
+          this.accessibilityManager?.enable();
+          this.div.style.display = "";
+          this.scale = scale;
+          this.show();
+        },
+        function (reason) {
+          // Cancelled or failed to render text layer; skipping errors.
+        }
+      );
+    }
+  }
+
+  #setDimensions(viewport) {
+    const { div } = this;
+    const [pageLLx, pageLLy, pageURx, pageURy] = viewport.viewBox;
+    const pageWidth = pageURx - pageLLx;
+    const pageHeight = pageURy - pageLLy;
+    const { style } = div;
+
+    style.width = `calc(var(--scale-factor) * ${pageWidth}px)`;
+    style.height = `calc(var(--scale-factor) * ${pageHeight}px)`;
+  }
+
+  #update(viewport) {
+    this.div.setAttribute("data-main-rotation", viewport.rotation);
+  }
+
+  hide() {
+    this.div.setAttribute("hidden", "true");
+  }
+
+  show() {
+    this.div.removeAttribute("hidden");
   }
```

Comment:
These methods probably need to be updated, given the CSS changes made previously.